src/physical/tabular/operations/triescan_minus.rs

```diff
@@ -1,14 +1,260 @@
+    subtract_current_layers: Vec<usize>,
+
+    /// Current layer of this scan.
+    current_layer: Option<usize>,
+
+    /// List of `ColumnScanSubtract` where every entry represents one level of the resulting trie.
+    column_scans: Vec<UnsafeCell<ColumnScanT<'a>>>,
+}
+
+/// Struct containing additional information for the setting up a [`TrieScanSubtract`] iterator.
+/// One of these structures should be associated with each trie that is to be subtracted.
+#[derive(Debug, Clone)]
+pub struct SubtractInfo {
+    /// Which layers of the main trie are subtracted.
+    /// For example, if we have a main trie with for layers: a(x, y, z, w)
+    /// and we want to subtract the trie associated with this struct b(y, w)
+    /// this vector would contain: [1, 3]
+    pub used_layers: Vec<usize>,
+}
+
+impl SubtractInfo {
+    /// Create new [`SubtractInfo`].
+    pub fn new(mut used_layers: Vec<usize>) -> Self {
+        used_layers.sort();
+
+        Self { used_layers }
+    }
+
+    /// Applies a permutation representing a reordering of the trie scan that this info is attached to.
+    pub fn apply_permutation(&mut self, permutation: Permutation) {
+        for layer in &mut self.used_layers {
+            *layer = permutation.get(*layer);
+        }
+        self.used_layers.sort();
+    }
+}
+
+impl<'a> TrieScanSubtract<'a> {
+    /// Construct new [`TrieScanSubtract`] object.
+    pub fn new(
+        trie_main: TrieScanEnum<'a>,
+        tries_subtract: Vec<TrieScanEnum<'a>>,
+        infos: Vec<SubtractInfo>,
+    ) -> Self {
+        debug_assert!(tries_subtract.len() == infos.len());
+        debug_assert!(tries_subtract
+            .iter()
+            .all(|trie_subtract| trie_main.get_types().len() >= trie_subtract.get_types().len()));
+        debug_assert!(infos
+            .iter()
+            .all(|info| trie_main.get_types().len() >= info.used_layers.len()));
+        debug_assert!(
+            tries_subtract
+                .iter()
+                .zip(infos.iter())
+                .all(|(trie_subtract, info)| info.used_layers.len()
+                    == trie_subtract.get_types().len())
+        );
+        debug_assert!(infos.iter().all(|info| info.used_layers.is_sorted()));
+        debug_assert!(infos.iter().all(|info| info
+            .used_layers
+            .iter()
+            .all(|&l| l < trie_main.get_types().len())));
+
+        let arity_main = trie_main.get_types().len();
+
+        let last_used_layers: Vec<usize> = infos
+            .iter()
+            .map(|info| {
+                *info
+                    .used_layers
+                    .iter()
+                    .max()
+                    .expect("There should be at least one layer that is used.")
+            })
+            .collect();
+        let mut column_scans = Vec::<UnsafeCell<ColumnScanT<'a>>>::with_capacity(arity_main);
+
+        for layer in 0..arity_main {
+            unsafe {
```

Comment:
Can we limit the scope of the `unsafe` for example by having an additional `let` expression before the `if`?